### PR TITLE
#159817309 Sandbox Environment for The Developers

### PR DESCRIPTION
**Pull request by WinstonKamau**
## Description (what problem you're fixing)

  - The developers need a local environment that mirrors staging.
 

## Fix (what you did to fix it)

  - I created part of a docker replica of the infrastructure that is on staging.

#### Step 1 Clone the frontend and backend repositories.
e.g.
```
mkdir soc-app
cd soc-app
git clone -b ft-sandbox-159817309 https://github.com/WinstonKamau/andela-societies-backend
git clone -b ft-sandbox-159817309 https://github.com/WinstonKamau/andela-societies-frontend
```
Ensure port 4021, 4022 and 4023 are not running on your machine.

#### Step 2 Set up the backend: Environment Variables
Change directory to the backend
```
cd andela-societies-backend
```
Ensure you have **ALL** your environment variables set in the src/.env file without any quotation marks. 
```
PRIVATE_KEY_TEST={some_value}
PUBLIC_KEY_TEST={some_value}
PUBLIC_KEY={some_value}
DEV_DATABASE={some_value}
DATABASE_URL={some_value}
APP_SETTINGS={some_value}
ANDELA_API_URL={some_value}
DEV_TOKEN={some_value}
MAIL_GUN_URL={some_value}
MAIL_GUN_API_KEY={some_value}
MAIL_GUN_TEST={some_value}
CELERY_BROKER_URL={some_value}
CELERY_BACKEND=rpc://
SENDER_CREDS={some_value}
CIO={some_value}
```

The value of these variables are obtained from the TTL, fellow developers in the team or the DevOps engineer in the team.

For the sandbox environment strictly:
* Do not place **export** before the variable assignment in the .env file as this application uses [dot-env](https://pypi.org/project/python-dotenv/)
e.g. This will not work as docker picks even the export name.
```
export PRIVATE_KEY_TEST=
export PUBLIC_KEY_TEST=
```
* Again do not assign quotation marks[""] to the any variable .
e.g
```
Wrong implementation:
DEV_TOKEN="ewogICIrKysrKysrKysrKyI6IHsKICAgICItLS0tLS0tLS0tIjogIj09PT09PT09PT09PSIsCiAgICAiLS0tLS0tLS0iOiAiwqLCosKiwqLCosKiwqLCosKiwqLCosKiIiwKICAgICItLS0tLS0tLS0iOiAiIyMjIyMjIyMjIyMjIyMiLAogICAgIi0tLS0tLS0tIjogIsKjwqPCo8KjwqPCo8KjwqPCo8KjwqPCo8KjwqPCo8KjIiwKICAgICItLS0tLS0tLSI6ICItLS0tLS0tLS0tLS0iLAogICAgIi0tLS0tLS0tLS0tLS0iOiAiZW1wdHkiLAogICAgIi0tLS0tLS0tLS0tLSI6IHsKICAgICAgIi0tLS0tLS0tLS0tIjogIi0tLS0tLS0tLS0tLS0tLS0iLAogICAgICAiLS0tLS0tLS0tIjogIi0tLS0tLS0tLS0tLS0tLS0tLS0tIgogICAgfQogICAgIioqKioqKioqKioiOiB7CiAgICAgICJub3RoaW5nIjogIm5vdGhpbmciLAogICAgICAiemlwIjogInppcCB6aXAgemlwIgp9CiAgfSwKICAiaWF0IjogMDAwMDAwMDAwMDAwLAogICJleHAiOiAwMDAwMDAwMDAwMDAwLAp9Cg=="

Correct implementation:
DEV_TOKEN=ewogICIrKysrKysrKysrKyI6IHsKICAgICItLS0tLS0tLS0tIjogIj09PT09PT09PT09PSIsCiAgICAiLS0tLS0tLS0iOiAiwqLCosKiwqLCosKiwqLCosKiwqLCosKiIiwKICAgICItLS0tLS0tLS0iOiAiIyMjIyMjIyMjIyMjIyMiLAogICAgIi0tLS0tLS0tIjogIsKjwqPCo8KjwqPCo8KjwqPCo8KjwqPCo8KjwqPCo8KjIiwKICAgICItLS0tLS0tLSI6ICItLS0tLS0tLS0tLS0iLAogICAgIi0tLS0tLS0tLS0tLS0iOiAiZW1wdHkiLAogICAgIi0tLS0tLS0tLS0tLSI6IHsKICAgICAgIi0tLS0tLS0tLS0tIjogIi0tLS0tLS0tLS0tLS0tLS0iLAogICAgICAiLS0tLS0tLS0tIjogIi0tLS0tLS0tLS0tLS0tLS0tLS0tIgogICAgfQogICAgIioqKioqKioqKioiOiB7CiAgICAgICJub3RoaW5nIjogIm5vdGhpbmciLAogICAgICAiemlwIjogInppcCB6aXAgemlwIgp9CiAgfSwKICAiaWF0IjogMDAwMDAwMDAwMDAwLAogICJleHAiOiAwMDAwMDAwMDAwMDAwLAp9Cg==
```

#### Step 3 Start the backend
Start up your docker daemon. Ensure you are on the root of the folder andela-societies-backend and run the command below to start the backend application:
```
make start
```
If running the sandbox for the **first** time seed the database.
```
make seed
```

#### Step 4 Start the frontend application
Change directory into the andela-societies-frontend application and run the command below:
```
cd andela-societies-frontend
make start
```

#### Step 5 Other commands for the sandbox
While on the root of the folder for the frontend you can run the command: 
* `make help`: To determine all the commands you can run on the frontend.
* `make stop`: To stop the frontend application and delete running containers.
* `make tear`: To destroy resources for the frontend application. *Sometimes the network may not be deleted if the backend is running. Just ignore the error*

While on the root of the folder for the backend you can run the command:
* `make stop`: To stop the backend application and delete running containers.
* `make tear`: To destroy resources for the backend application. *Sometimes the network may not be deleted if the frontend is running. Just ignore the error. Also while starting the application again after running this command you'll have to seed the database since the postgres database is deleted with this command.*
* `make seed`: To seed the database and link cohorts to societies. Should be done once after running `make start` for the first time or when a `make tear` has been run before and you desire to start the sandbox again.
* `make help`: To determine all the commands you can run on the backend. **Note:** *you'll meet a plethora of commands that do not relate to the sandbox.*

#### Step 6 View the database 
To view the database use the following credentials. (The database should be running.)
* host: 127.0.0.1
* user: soc
* password: soc
* database: soc-sandbox
* port: 4023

### CODE REVIEW

**Comment by georgreen** [comment]
@WinstonKamau Will you please take me through this? What does this change about local dev work?

**Comment by LarryWachira** [comment]
What wizardry is this docker? Finally got things working yesterday and I'm blown away! Must. Learn. Docker.
On a different note, would it be possible to have either makefile set up both the backend and frontend in one command?

**Comment by WinstonKamau** [comment]
@LarryWachira it is possible but there are two limitations that prevented us from doing so: 
*  It will limit the developer from viewing various branches.  For example, at the moment, after this is merged a backend developer will be able to checkout to any branch and deploy the changes to the branch, he has checked out to, to the sandbox.
* It will limit a developer from editing either the frontend or backend. For example, after the P.R is merged, a backend developer can easily go to the frontend repo (from wherever he had cloned it to) and edit it.  If we make it automatic that it will clone the frontend repo for him, it will limit him from changing the branch of the frontend, it will also be difficult to locate the frontend has been cloned to edit it.

**Comment by LarryWachira** [comment]
I think is ready for merging now...

**Comment by AnthonyGW** [comment]
After we get logs to output to a file, we should implement volume sharing so that devs always have access to these log files.

**Comment by WinstonKamau** [comment]
@AnthonyGW Changes to add the log files have been pushed. Please review.

**Comment by kimani-ndegwa** [comment]
@AnthonyGW Please attend to these conflicts

File: docker/dev/docker-compose.yml [hunk 1]
```diff
@@ -0,0 +1,42 @@
+version: '3.0'
+
+services:
+  backend:
+    build:
+      context: ../../
+      dockerfile: docker/dev/Dockerfile
+    image: soc-backend:sandbox
+    volumes:
```
Comment (AnthonyGW) [inline docker/dev/docker-compose.yml]: I think the devs should be informed that they need an existing logs file in the parent-parent directory. Or is one created for them?

Comment (WinstonKamau) [reply docker/dev/docker-compose.yml]: After testing it, I have found out that the logs folder is made automatically by docker itself.

Comment (AnthonyGW) [reply docker/dev/docker-compose.yml]: Cool

**Comment by WinstonKamau** [comment]
@kimani-ndegwa The conflicts have been fixed. This is ready for merging.

**Comment by LarryWachira** [comment]
@zachang Consider trying out this as well. It makes setting up everything much easier.

**Pull request merged**